platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/RelevantProjectsRegistry.kt

```diff
@@ -16,24 +16,27 @@
 
 package org.gradle.internal.cc.impl
 
+import com.google.common.collect.Sets.newConcurrentHashSet
 import org.gradle.api.internal.artifacts.configurations.ProjectComponentObservationListener
 import org.gradle.api.internal.project.ProjectState
 import org.gradle.api.internal.project.ProjectStateRegistry
 import org.gradle.execution.plan.Node
 import org.gradle.internal.build.BuildState
+import org.gradle.internal.service.scopes.ParallelListener
 import org.gradle.internal.service.scopes.Scope
 import org.gradle.internal.service.scopes.ServiceScope
 import org.gradle.util.Path
 
 
+@ParallelListener
 @ServiceScope(Scope.Build::class)
 class RelevantProjectsRegistry(
     private val build: BuildState,
     private val projectStateRegistry: ProjectStateRegistry
 ) : ProjectComponentObservationListener {
 
     private
-    val targetProjects = mutableSetOf<ProjectState>()
+    val targetProjects = newConcurrentHashSet<ProjectState>()
```

Comment:
💭  Given the observed contention and the migration to parallel listeners, it's safe to assume the `projectObserved` method will be called concurrently.

Now, its implementation tries to look up the project state, which is another project registry lock. We can avoid potential contention on it by not looking up the state immediately, and instead only collect the target paths. All look-ups can be done later when the `relevantProjects` is called.